CONSOLE-2380: Add UI for enabling/disabling dynamic plugins during op…

rhamilto opened the pull request:
…erator install

Test CatalogSource image url: `quay.io/spadgett/example-registry:latest`
Test operator from above CatalogSource:  `etcd`

**Install Operator form**
If the operator is not trusted, the value defaults to `Disabled` and a warning appears when changing to `Enabled`
<img width="1498" alt="Screen Shot 2021-02-17 at 9 50 46 AM" src="https://user-images.githubusercontent.com/895728/108221430-c5aa9c80-7105-11eb-8220-1869f0bef184.png">
<img width="1381" alt="Screen Shot 2021-02-17 at 9 51 01 AM" src="https://user-images.githubusercontent.com/895728/108221433-c6dbc980-7105-11eb-88c5-a4117e82f0ee.png">

**Installed Operators list**
If an operator has a plugin available and it is not enabled, a link with popover is added to the status column
<img width="1339" alt="Screen Shot 2021-02-16 at 10 15 52 AM" src="https://user-images.githubusercontent.com/895728/108082824-7babb300-7040-11eb-8c56-99bf0a228979.png">

**Operator Details**
<img width="1444" alt="Screen Shot 2021-02-17 at 9 57 03 AM" src="https://user-images.githubusercontent.com/895728/108222273-9fd1c780-7106-11eb-9455-4c395c7b1894.png">

If the operator is not trusted, the value defaults to `Disabled` and a warning appears when changing to `Enabled`
<img width="1469" alt="Screen Shot 2021-02-17 at 9 57 14 AM" src="https://user-images.githubusercontent.com/895728/108222378-bd069600-7106-11eb-8cc1-786bf73f78ab.png">
<img width="1472" alt="Screen Shot 2021-02-17 at 9 57 24 AM" src="https://user-images.githubusercontent.com/895728/108222380-bd9f2c80-7106-11eb-96c8-3e1b5cb29afa.png">

If there are more than one plugins associated with the operator, display the operator name to delineate:
<img width="239" alt="Screen Shot 2021-02-17 at 11 02 50 AM" src="https://user-images.githubusercontent.com/895728/108231608-ec6dd080-710f-11eb-8370-a4fe5127887e.png">
<img width="640" alt="Screen Shot 2021-02-17 at 10 44 20 AM" src="https://user-images.githubusercontent.com/895728/108229144-87b17680-710d-11eb-8618-bba776bc2082.png">
<img width="191" alt="Screen Shot 2021-02-17 at 10 50 44 AM" src="https://user-images.githubusercontent.com/895728/108229711-0e665380-710e-11eb-85fb-ee67d676bd0d.png">

rhamilto commented:
/retest

rhamilto commented:
/retest

spadgett commented on frontend/packages/operator-lifecycle-manager/src/components/operator-hub/operator-hub-subscribe.tsx:
```diff
@@ -334,6 +373,16 @@ export const OperatorHubSubscribeForm: React.FC<OperatorHubSubscribeFormProps> =
+          {
+            path: '/spec/plugins',
+            value: _.union(previousPlugins, enabledPlugins),
+            op: 'add',
```
I'm debating if we need to remove existing enabled plugins if we are reinstalling an operator and the user picks "Disabled."

spadgett replied:
> I'm debating if we need to remove existing enabled plugins if we are reinstalling an operator and the user picks "Disabled."

I think the logic for this is pretty simple and we should go ahead and do it. Something like

```ts
value: [
  ...previousPlugins.filter((plugin: string) => !csvPlugins.includes(plugin)),
  ...enabledPlugins,
],
```

rhamilto replied:
This does not work as anticipated since `op: 'add'`.  I think we need to `op: 'remove'` as a separate condition?

spadgett replied:
Hm, what happens? I'd think we do want add, which works as either add or replace depending on whether the array exists. Here we're replacing the entire array. But I haven't tested.

rhamilto replied:
I think https://github.com/openshift/console/pull/8175/commits/f4db0cb36e51b77e875f3a77d4b3c262eee52aeb is what we want, @spadgett?

rhamilto replied:
Updated in be8d619ae2a06f013d786446acfe0fd491433c35

spadgett commented on frontend/packages/operator-lifecycle-manager/src/const.ts:
```diff
@@ -8,3 +8,4 @@ export const OPERATOR_TYPE_ANNOTATION = 'operators.operatorframework.io/operator
 export const NON_STANDALONE_ANNOTATION_VALUE = 'non-standalone';
 export const INTERNAL_OBJECTS_ANNOTATION = 'operators.operatorframework.io/internal-objects';
 export const DEFAULT_SOURCE_NAMESPACE = 'openshift-marketplace';
+export const CLUSTER_SERVICE_VERSION_PLUGINS_ANNOTATION = 'console.openshift.io/plugins';
```
We might call this `OPERATOR_PLUGINS_ANNOTATION` to match the CSV annotation names above.

spadgett commented on frontend/packages/operator-lifecycle-manager/src/utils/consolePluginFormGroup.tsx:
```diff
@@ -0,0 +1,67 @@
+                enabled={enabledPlugins.includes(plugin)}
+                onChange={(enabled: boolean) => setPluginEnabled(plugin, enabled)}
+              />
+              <ConsolePluginWarning
```
I'd suggest we have one warning for all plugins. (You'd just need to check `enabled={enabledPlugins?.length}` in that case.

spadgett reviewed (commented):
Thanks @rhamilto

rhamilto commented:
@spadgett, I believe I've addressed all your comments with the exception of https://github.com/openshift/console/pull/8175#discussion_r577741880 since you are debating it.  See https://github.com/openshift/console/pull/8175/commits/e597d1e258bd2e037a2e1231d58bf2d151c4a3da

rhamilto commented:
/hold for approvals
/assign @ahardin-rh @yapei @sferich888

spadgett commented on frontend/packages/operator-lifecycle-manager/src/const.ts:
```diff
@@ -8,3 +8,5 @@ export const OPERATOR_TYPE_ANNOTATION = 'operators.operatorframework.io/operator
 export const INTERNAL_OBJECTS_ANNOTATION = 'operators.operatorframework.io/internal-objects';
 export const DEFAULT_SOURCE_NAMESPACE = 'openshift-marketplace';
+export const OPERATOR_PLUGINS_ANNOTATION = 'console.openshift.io/plugins';
+export const CONSOLE_OPERATOR_CONFIG_NAME = 'cluster';
```
We might move this elsewhere since it's not specific to OLM. Maybe

https://github.com/openshift/console/blob/master/frontend/packages/console-shared/src/constants/resource.ts

spadgett commented on frontend/packages/operator-lifecycle-manager/src/utils.ts:
```diff
@@ -1,6 +1,44 @@
+import { K8sResourceKind, ObjectMetadata, Patch } from '@console/internal/module/k8s';
+import { INTERNAL_OBJECTS_ANNOTATION, OPERATOR_PLUGINS_ANNOTATION } from './const';
+
+export const getCatalogSourceIsTrusted = (catalogSource: string): boolean =>
```
I'd suggest

```suggestion
export const isCatalogSourceTrusted = (catalogSource: string): boolean =>
```

spadgett commented on frontend/packages/operator-lifecycle-manager/src/utils/consolePluginFormGroup.tsx:
```diff
@@ -0,0 +1,67 @@
+import * as React from 'react';
```
nit: In this package, we've been kebab casing the file names (e.g., console-plugin-form-group.tsx)

spadgett commented:
Looks good overall, just a few small comments. Thanks @rhamilto !

rhamilto commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -251,6 +263,106 @@ const ManagedNamespaces: React.FC<ManagedNamespacesProps> = ({ obj }) => {
+          {csvPlugins.map((plugin) => (
+            <dd key={plugin} className="co-clusterserviceversion-details__field-description">
+              {csvPluginsCount > 1 && (
+                <strong className="text-muted">{t('olm~{{plugin}}:', { plugin })} </strong>
```
Even though it includes a `:`?

rhamilto replied:
That is, does it make sense to externalize just the colon without the context?

spadgett replied:
Hm, you might be right that the punctuation is slightly different in different languages. We can leave it.

rhamilto commented on frontend/packages/operator-lifecycle-manager/src/components/modals/console-plugin-modal.tsx:
```diff
@@ -0,0 +1,81 @@
+    <form onSubmit={submit} name="form" className="modal-content">
+      <ModalTitle>
+        {t('olm~Console UI extension')}
+        {csvPluginsCount > 1 && <>{t('olm~ - {{plugin}}', { plugin })}</>}
```
Even though it includes a `-`?

spadgett replied:
Hm, maybe we want two separate messages:

```
olm~Console UI extension
```

and

```
olm~Console UI extension - {{plugin}}
```

That way the translators have the full context if we think they might need to change punctuation.

yapei commented:
/label qe-approved

ahardin-rh commented:
/label docs-approved

spadgett commented on frontend/packages/operator-lifecycle-manager/src/components/operator-hub/operator-hub-subscribe.tsx:
```diff
@@ -334,6 +373,23 @@ export const OperatorHubSubscribeForm: React.FC<OperatorHubSubscribeFormProps> =
+        await k8sPatch(ConsoleOperatorConfigModel, consoleOperatorConfig, [
+          {
+            path: '/spec/plugins',
+            value: [
+              ...previousPlugins.filter((plugin: string) => !csvPlugins.includes(plugin)),
+              ...enabledPlugins,
+            ],
```
```suggestion
            value: updatedPlugins,
```

rhamilto replied:
Done

spadgett commented on frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx:
```diff
@@ -251,6 +260,106 @@ const ManagedNamespaces: React.FC<ManagedNamespacesProps> = ({ obj }) => {
+  });
+  const aPluginIsDisabled =
+    !consoleOperatorConfig?.spec?.plugins?.length ||
+    csvPlugins.some((plugin) => isPluginEnabled(consoleOperatorConfig, plugin) === false);
```
Can this simply be...?

```suggestion
    csvPlugins.some((plugin) => !isPluginEnabled(consoleOperatorConfig, plugin));
```

I assume `isPluginEnabled` always returns a boolean and never null.

rhamilto replied:
Yes.  Good catch.

spadgett reviewed (commented):
Thanks, @rhamilto. Looks good

rhamilto commented:
Thanks, @spadgett.  Updated, rebased, squashed.

rhamilto commented:
/retest

rhamilto commented:
/retest

rhamilto commented:
/retest

rhamilto commented:
/retest

rhamilto commented:
/retest

spadgett reviewed (approved):
/lgtm

sferich888 commented:
/label px-approved
/unassign @sferich888

spadgett commented:
/hold cancel

rhamilto commented:
/retest

spadgett commented:
The failing test is removed in #8145. Let's wait for that PR.

/hold

itsptk commented:
@spadgett You mentioned today that these resources are really called "console plugins." I was wondering if you thought the "Console UI extension" term used in this UI was perhaps ambiguous? Maybe it really should have been "Console plugin" or maybe even a mix of actual resource and its purpose, like "Console UI plugin"? cc @alimobrem

spadgett commented:
/hold cancel
